DDD-1.md

```diff
@@ -0,0 +1,179 @@
+
+So far, the following classes have been identified:
+
+1. `ChangeEventSourceCoordinator`
+2. `RelationalSnapshotChangeEventSource`
+3. `AbstractSnapshotChangeEventSource`
+
+#### Example of a change in the `ChangeEventSourceCoordinator` logic
+
+In partitioned mode, the coordinator, instead of just switching from snapshot to the streaming sources, will iterate over a collection of partitions/offsets and reuse each source for each of them:
+
+**Nonpartitioned (current logic)**:
+
+```java
+// offset context is part of the source state
+snapshotSource  = new SnapshotSource(offsetContext)
+streamingSource = new StreamingSource()
+
+// snapshot is executed for its only partition
+result = snapshotSource.execute()
+
+// streaming is done from the only partition/offset
+streamingSource.execute(result.getOffset())
+```
+
+**Partitioned**:
+
+```java
+// offset context is part of the interface, not the state
+snapshotSource  = new PartitionedSnapshotSource()
+streamingSource = new PartitionedStreamingSource()
+
+taskContex.forEach((partition, offset) -> {
+  // snapshot is executed for each given partition
+  result = snapshotSource.execute(partition)
+  // result are stored per-partition
+  results.put(partition, result);
+})
+
+results.forEach((partition, offset) -> {
+  // streaming is done from one partition at a time
+  streamingSource.execute(partition, result.getOffset())
+})
+```
+
+### Move control loop from streaming sources to the coordinator
+
+The above iteration of streaming sources won't work if implemented as is. The reason is that `streamingSource.execute()` will block until the task is stopped.
+
+We need to remove `while (context.isRunning())` from the logic of partitioned streaming sources (e.g. SQL Server) and move this logic to the coordinator:
+
+```java
+while (context.isRunning()) {
+  results.forEach((partition, offset) -> {
+    // streaming is done per-partition
+    streamingSource.execute(partition, result.getOffset())
+  })
+}
+```
+
+### Changes in mapping table identifiers to topic names and internal schema representation
+
+In order to accomodate changes from multiple databases, instead of naming topics as `<server>.<schema>.<table>`, the connector will name them as `<server>.<database>.<schema>.<table>`.
+
+The changes implemented in [DBZ-1089](https://issues.redhat.com/browse/DBZ-1089) should be partially reverted since for SQL Server both the database name and schema name should be taken into account when building schema.
+
+## TODO:
+
+1. Define configuration for multiple databases to be processed by a connector. Only one of `database.dbname` and this one must be configured on a given connector.
+2. Provide the upgrade path. The addition of the database name to the records' source partition will make it impossible for the upgraded connector to resume from the offset produced by the previous version. Likely, we'll need to keep the "legacy" version and let the consumers migrate their data and switch over.
```

Comment:
Why do you think would it be impossible for the connector to read back the old offset format?

Reply:
No reason to think it won't work. I like the approach.

Reply:
Ok, cool. Let's do this then.